File: .agents/skills/desktop-upgrade-harness/scripts/harness_upgrade_preflight.py
"""Harness 升级写入前的快照与计划复验。"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Any

from harness_upgrade_core import (
    Snapshot,
    UpgradeError,
    assert_safe_path,
    canonical_directory,
    require_git_root,
    safe_relative_path,
    snapshot_file,
    validate_snapshot,
)


def assert_current_snapshot(
    root: Path,
    relative: str,
    expected: Snapshot | None,
    label: str,
) -> Path:
    """重验普通文件快照和祖先边界，阻止计划后的静默变化。"""

    path = assert_safe_path(
        root,
        root / relative,
        label,
        final_may_be_missing=expected is None,
    )
    if expected is None:
        if os.path.lexists(path):
            raise UpgradeError(f"{label}意外存在：{relative}")
        return path
    if not path.is_file() or path.is_symlink():
        raise UpgradeError(f"{label}不是普通文件：{relative}")
    observed = snapshot_file(path)
    if observed != expected:
        raise UpgradeError(
            f"{label}在 plan 后发生变化：{relative}；"
            f"预期 {expected}，实际 {observed}"
        )
    return path


def assert_plan_still_current(plan: dict[str, Any]) -> None:
    """在写入前重验控制文件、Git 身份以及计划覆盖的两棵文件树。"""

    candidate = canonical_directory(Path(plan["candidate_root"]), "候选根目录")
    target = canonical_directory(Path(plan["target_root"]), "目标根目录")
    ownership = Path(plan["ownership_path"])
    if snapshot_file(ownership) != validate_snapshot(
        plan["ownership_snapshot"],
        "ownership_snapshot",
    ):
        raise UpgradeError("所有权 manifest 在 plan 后发生变化")
    lock_path = Path(plan["lock_path"])
    expected_lock = validate_snapshot(plan["lock_snapshot"], "lock_snapshot")
    if expected_lock is None:
        if os.path.lexists(lock_path):
            raise UpgradeError("上游 lock 在 plan 后出现")
    elif snapshot_file(lock_path) != expected_lock:
        raise UpgradeError("上游 lock 在 plan 后发生变化")
    if require_git_root(target) != plan["target_git"]:
        raise UpgradeError("目标 Git 身份或工作树状态在 plan 后发生变化")
    for item in plan["actions"]:
        relative = safe_relative_path(item["path"])
        assert_current_snapshot(
            candidate,
            relative,
            validate_snapshot(item.get("candidate"), f"{relative}.candidate"),
            "候选",
        )
        assert_current_snapshot(
            target,
            relative,
            validate_snapshot(item.get("target"), f"{relative}.target"),
            "目标",
        )

File: .agents/skills/desktop-upgrade-harness/scripts/harness_upgrade_preflight.mjs
/** Harness 升级写入前的快照与计划复验。 */

import fs from "node:fs";
import path from "node:path";
import { isDeepStrictEqual } from "node:util";
import { UpgradeError, assertSafePath, canonicalDirectory, requireGitRoot, safeRelativePath, snapshotFile, validateSnapshot } from "./harness_upgrade_safety.mjs";

function lexists(value) { try { fs.lstatSync(value); return true; } catch (error) { if (error?.code === "ENOENT") return false; throw error; } }

export function assertCurrentSnapshot(root, relative, expected, label) {
  const target = assertSafePath(root, path.join(root, ...relative.split("/")), label, { finalMayBeMissing: expected === null });
  if (expected === null) { if (lexists(target)) throw new UpgradeError(`${label}意外存在：${relative}`); return target; }
  const stat = lexists(target) ? fs.lstatSync(target) : null;
  if (!stat?.isFile() || stat.isSymbolicLink()) throw new UpgradeError(`${label}不是普通文件：${relative}`);
  const observed = snapshotFile(target);
  if (!isDeepStrictEqual(observed, expected)) throw new UpgradeError(`${label}在 plan 后发生变化：${relative}；预期 ${JSON.stringify(expected)}，实际 ${JSON.stringify(observed)}`);
  return target;
}

export function assertPlanStillCurrent(plan) {
  const candidate = canonicalDirectory(plan.candidate_root, "候选根目录");
  const target = canonicalDirectory(plan.target_root, "目标根目录");
  if (!isDeepStrictEqual(snapshotFile(plan.ownership_path), validateSnapshot(plan.ownership_snapshot, "ownership_snapshot"))) throw new UpgradeError("所有权 manifest 在 plan 后发生变化");
  const expectedLock = validateSnapshot(plan.lock_snapshot, "lock_snapshot");
  if (expectedLock === null) { if (lexists(plan.lock_path)) throw new UpgradeError("上游 lock 在 plan 后出现"); }
  else if (!isDeepStrictEqual(snapshotFile(plan.lock_path), expectedLock)) throw new UpgradeError("上游 lock 在 plan 后发生变化");
  if (!isDeepStrictEqual(requireGitRoot(target), plan.target_git)) throw new UpgradeError("目标 Git 身份或工作树状态在 plan 后发生变化");
  for (const item of plan.actions) {
    const relative = safeRelativePath(item.path);
    assertCurrentSnapshot(candidate, relative, validateSnapshot(item.candidate, `${relative}.candidate`), "候选");
    assertCurrentSnapshot(target, relative, validateSnapshot(item.target, `${relative}.target`), "目标");
  }
}
